include/picongpu/particles/fusion/detail/Creation.hpp

```diff
@@ -0,0 +1,455 @@
+/* Copyright 2025-2025 Filip Optolowicz
+ *
+ * This file is part of PIConGPU.
+ *
+ * PIConGPU is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * PIConGPU is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+// PIConGPU Includes
+#include "picongpu/defines.hpp"
+#include "picongpu/particles/fusion/param.hpp"
+#include "picongpu/traits/frame/GetCharge.hpp"
+#include "picongpu/traits/frame/GetMass.hpp"
+#include "picongpu/unitless/simulation.unitless"
+
+// PMacc Includes
+#include <pmacc/lockstep.hpp>
+#include <pmacc/mappings/kernel/AreaMapping.hpp>
+#include <pmacc/math/Vector.hpp>
+#include <pmacc/math/operation.hpp>
+#include <pmacc/memory/shared/Allocate.hpp>
+#include <pmacc/mpi/MPIReduce.hpp>
+#include <pmacc/mpi/reduceMethods/Reduce.hpp>
+#include <pmacc/particles/algorithm/ForEach.hpp>
+#include <pmacc/random/RNGProvider.hpp>
+#include <pmacc/random/distributions/Uniform.hpp>
+
+// Standard Library Includes
+#include <cmath>
+#include <cstddef>
+#include <limits>
+
+#if (BOOST_LANG_CUDA || BOOST_COMP_HIP)
+#    include <mallocMC/mallocMC.hpp>
+#endif
+
+namespace picongpu::particles::fusion
+{
+    namespace detail
+    {
+        using namespace precision;
+
+        /**
+         * @brief constexpr implementation used instead of std::round
+         * @param x value to round
+         * @return rounded value
+         */
+        template<typename T>
+        constexpr T round_cx(T x)
+        {
+            T truncated = static_cast<long long>(x);
+            T fraction = x - truncated;
+
+            if(fraction >= 0.5)
+            {
+                return truncated + 1.0;
+            }
+            if(fraction <= -0.5)
+            {
+                return truncated - 1.0;
+            }
+            return truncated;
+        }
+
+        // Cap-aware validator allowing W > 1 up to caps
+        constexpr bool areWeightsValidWithCaps(
+            float_COLL W1,
+            float_COLL W2,
+            float_COLL W3,
+            float_COLL W4,
+            float_COLL c3,
+            float_COLL c4)
+        {
+            // clang-format off
+            return (W1 >= 0.0_X && W1 <= c3 &&
```

Comment:
`0.0_X` is not nessesary `float_COLL` you should always write `float_COLL{0.0}` to keep the precision.